lib/gears/reactive.lua

```diff
@@ -0,0 +1,647 @@
+--
+-- That design has one big limitation. It cannot detect any changes which are
+-- not directly part of `gears.object` instance. You cannot use random tables
+-- and expect the function to be called when it's content change. To work
+-- around this, it is recommanded to make "hub" objects to store the data used
+-- within the reactive function:
+--
+-- Recommanded usage
+-- =================
+--
+-- The best way to use `gears.reactive` is when the value used within the
+-- expressions are part of other objects. It can be a `gears.watcher`, but
+-- it can also be a custom object:
+--
+-- @DOC_wibox_widget_declarative_reactive_EXAMPLE@
+--
+-- Limitations
+-- ===========
+--
+-- `gears.reactive` is pushing Lua way beyond what it has been designed for.
+-- Because of this, there is some limitations.
+--
+-- * This only works with `gears.object` and Core API objects which implement
+--   the `property::*****` signals correctly. If it is a regular Lua table
+--   or the property signals are incorrectly used, the value changes cannot
+--   be detected. If you find something that should work, but doesn't in
+--   one of the AwesomeWM API, please report a bug.
+-- * More generally, when making a custom `gears.object` with custom setters,
+--   it is the programmer responsibility to emit the signal. It is also
+--   required to only emit those signals when the property actually changes to
+--   avoid an unecessary performance loss.
+-- * Changing upvalue types after the reactive expression has been created wont
+--   be detected and will cause errors.
+-- * Internally, the engine tries its best to prevent the internal proxy objects
+--   to leak out the sandbox. However this cannot be perfect, at least not
+--   without adding limitation elsewhere. It is probably worth reporting a bug if
+--   you encounter such an issue. But set your expectations, not all corner case
+--   can be fixed.
+-- * Don't use rawset in the expression.
+-- * If the return value is a table, only the first 3 nesting levels are sanitized.
+--   Avoid using nested tables in the returned value if you can. `gears.object`
+--   instances *should* be fine.
+-- * This module will **NOT** try to track the change of other
+--   functions and methods called by the expression. It is **NOT** recursive
+--   and only the top level properties are tracked. This is a feature, not a
+--   bug. If it was recusrsive, this list of limitations or gotchas would be
+--   endless.
+-- * There is currently no way to disable a reactive expression once it's
+--   been defined. This may change eventually.
+-- * Rio Lua 5.1 (not LuaJIT 2.1) is currently not support. If you need it,
+--   report a bug.
+--
+-- @author Emmanuel Lepage-Vallee &lt;elv1313@gmail.com&gt;
+-- @copyright 2017-2020 Emmanuel Lepage-Vallee
+-- @classmod gears.reactive
+
+-- This file is provided under the BSD 2 clause license since it is better than
+-- any existing implementation (most of which only support Lua 5.1).
```

Comment:
Please also license it under the same license as AwesomeWM.

I do not know how these license combinations really work (IANAL), but introducing code with a different license sounds more complicated.

Reply:
BSD 2 clause code can be used in GPL2 code as-is, the dual license isn't required. The GPL virality converts the resulting aggregate from MIT or BSD2 to GPL2. The number of BSD clause is important, BSD4 is incompatible with the GPL (it requires to add the extra clause to your copy of the GPL2, which requires to get every copyright holder to agree. The GPL3 forbids extra clauses unless you rename the license). (IANAL, but have been working with them for 15 years with all the FOSS stuff I did over the years).